### ISSUE TITLE
Dynamic imports CORS errors

### ISSUE THREAD

**Issue by rijk** [meteor/meteor#10384]
I found similar issues here: #9888, here: #9747, and here: #9952. Those should be fixed. However, I'm running the latest version of everything (Meteor 1.8.0.1, dynamic-import 0.5.1), and still experiencing this issue.

Error:
```
[Error] Cross-origin redirection to https://www.envoy.rocks/__meteor__/dynamic-import/fetch denied by Cross-Origin Resource Sharing policy: Origin https://try.envoy.rocks is not allowed by Access-Control-Allow-Origin.
[Error] Fetch API cannot load https://envoy.rocks/__meteor__/dynamic-import/fetch due to access control checks.
[Error] Failed to load resource: Cross-origin redirection to https://www.envoy.rocks/__meteor__/dynamic-import/fetch denied by Cross-Origin Resource Sharing policy: Origin https://try.envoy.rocks is not allowed by Access-Control-Allow-Origin. (fetch, line 0)
[Error] Unhandled Promise Rejection: TypeError: Cross-origin redirection to https://www.envoy.rocks/__meteor__/dynamic-import/fetch denied by Cross-Origin Resource Sharing policy: Origin https://try.envoy.rocks is not allowed by Access-Control-Allow-Origin.
	t (62d32c0cf3a54438be4c7a1feea7ad9d97eb6cd3.js:21:915)
	(anonymous function) (62d32c0cf3a54438be4c7a1feea7ad9d97eb6cd3.js:1:8221)
	promiseReactionJob
[Error] Unhandled Promise Rejection: Error: Cannot find module '/imports/ui/components/panels/Timeline.jsx'
	t (62d32c0cf3a54438be4c7a1feea7ad9d97eb6cd3.js:21:915)
	(anonymous function) (62d32c0cf3a54438be4c7a1feea7ad9d97eb6cd3.js:1:8221)
	promiseReactionJob
```

![image](https://user-images.githubusercontent.com/159500/50219248-e85a9280-038e-11e9-87e0-bf1066f4ca4c.png)

Any idea what is going on? It used to work but it just broke all of a sudden. Maybe something changed on the Galaxy side? I am not sure if it started breaking right after my recent deploy, but it was a while since I did a deploy before that (May).

**Comment by aliogaili** [comment]
I see you've `try.envoy.rocks` and `www.envoy.rocks` these are considered cross origin requests so maybe that's why it's failing.

**Comment by aliogaili** [comment]
I think it's the ROOT_URL environment variable, from [here](https://docs.meteor.com/environment-variables.html#ROOT-URL). But if you changed your domain or settings.json this the mismatch might cause the problem, no?

**Comment by rijk** [comment]
I reverted to my May deploy now via the Galaxy UI, at least that is working.

**Comment by benjamn** [comment]
Is the https://envoy.rocks/__meteor__/dynamic-import/fetch URL getting redirected to https://try.envoy.rocks/__meteor__/dynamic-import/fetch somehow? Do you need to whitelist both in CORS?

**Comment by jamesmillerburgess** [comment]
We had a similar issue with our Galaxy deployment, and after talking to MDG support we fixed it by setting the `ROOT_URL` environment variable to the same domain at which we are advertising our site.

**Comment by rijk** [comment]
Ok, I finally had some more time to investigate. It starts breaking when I update from Meteor 1.6.1 to 1.8 (dynamic-import@0.3.0 -> dynamic-import@0.5.0).

@aliogaili , I am in fact not setting the `ROOT_URL` variable; my app runs on a wildcard domain (`*.envoy.rocks`) so I think that's why, but it was not necessary in the past either.

I tried to update to METEOR@1.8.1-beta.9 as well (as c4b5707747ce3c5bf339b62940da84628680a1d3 promises to solve the issue), but it did not work unfortunately (see https://try.envoy.rocks/drop/8Pz/gr8).

**Comment by rijk** [comment]
@benjamn sorry, Github only showed me your reply after I submitted mine because I had the tab open (no reactivity here 😄). 

I deployed now twice with `"ROOT_URL": "https://www.envoy.rocks"` (see screenshot) and `"ROOT_URL": "https://envoy.rocks"`. Neither solved the issue unfortunately.

<img width="536" alt="screenshot 2018-12-27 at 14 13 34" src="https://user-images.githubusercontent.com/159500/50481430-b88c3a00-09e1-11e9-8cee-2b2d6158f6f1.png">

**Comment by rijk** [comment]
Follow up — with Meteor 1.6.1 (working):

<img width="1356" alt="screenshot 2018-12-28 at 15 40 28" src="https://user-images.githubusercontent.com/159500/50518781-c7ddb700-0ab7-11e9-9e67-22b8dd5719e4.png">

With Meteor 1.8.1-beta.9 (not working):

<img width="1356" alt="screenshot 2018-12-28 at 15 44 21" src="https://user-images.githubusercontent.com/159500/50518790-d4faa600-0ab7-11e9-8ed7-6c0ab6cd3404.png">

Note the Request URL, that in 1.8.1 seems to be the generic URL (www.envoy.rocks). This isn't configured anywhere (I am not using `ROOT_URL`, and the Galaxy app domain I'm deploying to is envoy.rocks, not www.envoy.rocks) so I'm not sure where it's coming from.

**Comment by rijk** [comment]
@benjamn [This change](https://github.com/meteor/meteor/commit/a646bdbbd4b988f1f25c86159972fb1b70c13ce9) appears to be the culprit:

<img width="1016" alt="screenshot 2018-12-28 at 16 01 10" src="https://user-images.githubusercontent.com/159500/50519147-db8a1d00-0ab9-11e9-999a-29cd569b1b8c.png">

**Comment by rijk** [comment]
@benjamn hope I'm not bothering you if you're on vacation or something, but could you take a look at this?

**Comment by scarbies** [comment]
This is definitely an issue. I just upgraded from 1.6.0.1 to 1.8.1 and the identical CORS issue is preventing me from doing any dynamic imports on the client (used to work). I am accessing the same deployment from multiple subdomains so cannot set ROOT_URL for "each" one. This is a major limitation in this scenario.

**Comment by rijk** [comment]
I worked around it by forking the `dynamic-import` package locally.

**Comment by skarborg** [comment]
Thanks I might need to. Did you just roll back the change you referenced and viola?

**Comment by rijk** [comment]
This was (and is) the only changed compared to master:

![image](https://user-images.githubusercontent.com/159500/59018236-ae4e5180-8845-11e9-963d-600c73396a0d.png)

**Comment by Rijk** [comment]
Please do not close, this is still an issue.

**Comment by filipenevola** [comment]
thanks @rijk, I already labeled as `confirmed` then stale bot will ignore it.

**Comment by leonardoventurini** [comment]
@filipenevola looks like we are facing this issue as well... how likely would Meteor be to accept a Pull Request using the `location.origin` instead of the `ROOT_URL` for `dynamic-import`? A quite simple fix IMHO.

**Comment by Rijk** [comment]
@leonardoventurini that is in essence reverting the change that started all this: a646bdbbd4b988f1f25c86159972fb1b70c13ce9.

However, we need to investigate (and ideally get @benjamn input) why this change was made in the first place, to prevent regressions of course. It seems to have been related to CORS.

**Comment by iBeACoder** [comment]
We are also facing this issue since the app got bigger, and we switched over to dynamic imports. On iPhones, the website does not load on anything but Safari. Firefox and Opera do not support the site at all on iPhones.

**Comment by filipenevola** [comment]
@leonardoventurini I believe we could accept this PR if we enable this conditionally so the affected projects could use this solution. Are you willing to write this PR?

We should use the settings to enable/disable this behavior, in a key like:
```json
{
  "packages": {
    "dynamic-import": {
      "useLocationOrigin": true
    }
  }
}
```

**Comment by rijk** [comment]
@filipenevola stale bot ignoring doesn't seem to work :)
> I already labeled as confirmed then stale bot will ignore it.

Anyway, this is fixed in `dynamic-imports@0.5.3` with #11105, simply upgrade and add the following to your settings.json:

```json
{
  "public": {
    "packages": {
      "dynamic-import": {
        "useLocationOrigin": true
      }
    }
  }
}
```

**Comment by filipenevola** [comment]
I changed the label to Stage:Confirmed :)
